File: en/codes/kotlin/chapter_sorting/selection_sort.kt
/**
 * File: selection_sort.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_sorting

/* Selection sort */
fun selectionSort(nums: IntArray) {
    val n = nums.size
    // Outer loop: unsorted interval is [i, n-1]
    for (i in 0..<n - 1) {
        var k = i
        // Inner loop: find the smallest element within the unsorted interval
        for (j in i + 1..<n) {
            if (nums[j] < nums[k])
                k = j // Record the index of the smallest element
        }
        // Swap the smallest element with the first element of the unsorted interval
        val temp = nums[i]
        nums[i] = nums[k]
        nums[k] = temp
    }
}

/* Driver Code */
fun main() {
    val nums = intArrayOf(4, 1, 3, 1, 5, 2)
    selectionSort(nums)
    println("After selection sort, nums = ${nums.contentToString()}")
}

File: en/codes/go/chapter_sorting/selection_sort.go
// File: selection_sort.go
// Created Time: 2023-05-29
// Author: Reanon (793584285@qq.com)

package chapter_sorting

/* Selection sort */
func selectionSort(nums []int) {
	n := len(nums)
	// Outer loop: unsorted interval is [i, n-1]
	for i := 0; i < n-1; i++ {
		// Inner loop: find the smallest element within the unsorted interval
		k := i
		for j := i + 1; j < n; j++ {
			if nums[j] < nums[k] {
				// Record the index of the smallest element
				k = j
			}
		}
		// Swap the smallest element with the first element of the unsorted interval
		nums[i], nums[k] = nums[k], nums[i]

	}
}
